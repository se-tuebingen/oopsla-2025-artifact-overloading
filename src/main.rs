use overloading::compile_expr;
use overloading::syntax::parse_expr;

use clap::Parser;

#[derive(Parser)]
#[command(name = "overloading")]
struct Args {
    /// Input file to process
    filename: String,

    /// Use short output format
    #[arg(short, long)]
    short: bool,
}

/// Main entry point for the native CLI.
fn main() {
    let args = Args::parse();

    let content = std::fs::read_to_string(&args.filename).expect("Failed to read file");
    match parse_expr(&content) {
        Ok(expr) => {
            println!("{}", compile_expr(&expr, args.short));
        }
        Err(err) => {
            panic!("Failed to parse expression: {}", err);
        }
    }
}

/// Snapshot tests for the compiler examples.
#[cfg(test)]
#[test]
fn test_compiler_examples() {
    use insta::{assert_json_snapshot, glob, with_settings};

    glob!("../", "examples/*.over", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let output = match parse_expr(&input) {
            Ok(expr) => overloading::compile_expr_json(&expr, true),
            Err(err) => overloading::json::JsonProcessingResult::InferenceError {
                expression: input,
                error: err,
            },
        };

        // Use the filename (without extension) as snapshot suffix
        let filename = path.file_stem().unwrap().to_str().unwrap();
        with_settings!({
            snapshot_path => "../examples/snapshots",
            snapshot_suffix => filename,
            sort_maps => true,
        }, {
            assert_json_snapshot!(output);
        });
    });
}
