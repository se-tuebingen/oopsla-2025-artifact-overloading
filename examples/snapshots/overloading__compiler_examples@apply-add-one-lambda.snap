---
source: src/main.rs
expression: output
input_file: examples/apply-add-one-lambda.over
---
{
  "status": "success",
  "expression": "assume apply: ∀(A, B ). ((A) => B) => (A) => B\nassume addOneI: (Int) => Int\nassume addOneS: (String) => String\nassume addOneD: (Double) => Double\nlet addOne = a‹addOneI, addOneS, addOneD›\nλ x => apply(addOne)(x) => apply(addOne)(x)",
  "dimension_spans": "[spans]",
  "inferred_type": "(x_1) => app_5",
  "worlds": {
    "type": "many",
    "count": 3.0
  },
  "solutions": [
    [
      "a₁"
    ],
    [
      "a₃"
    ],
    [
      "a₂"
    ]
  ],
  "constraints": [
    {
      "from": "(Int) => Int",
      "choices": [
        "a₁"
      ],
      "to": "over_a_0"
    },
    {
      "from": "(String) => String",
      "choices": [
        "a₂"
      ],
      "to": "over_a_0"
    },
    {
      "from": "(Double) => Double",
      "choices": [
        "a₃"
      ],
      "to": "over_a_0"
    },
    {
      "from": "((A_2) => B_3) => (A_2) => B_3",
      "choices": [],
      "to": "(over_a_0) => app_4"
    },
    {
      "from": "app_4",
      "choices": [],
      "to": "(x_1) => app_5"
    }
  ],
  "bounds": [
    {
      "variable": "A_2",
      "lower": [],
      "upper": [
        {
          "type_name": "Double",
          "choices": [
            "a₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "a₂"
          ]
        }
      ]
    },
    {
      "variable": "B_3",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "a₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "a₂"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "app_5",
          "choices": []
        }
      ]
    },
    {
      "variable": "app_4",
      "lower": [
        {
          "type_name": "(A_2) => B_3",
          "choices": []
        }
      ],
      "upper": [
        {
          "type_name": "(x_1) => app_5",
          "choices": []
        }
      ]
    },
    {
      "variable": "app_5",
      "lower": [
        {
          "type_name": "Double",
          "choices": [
            "a₃"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "String",
          "choices": [
            "a₂"
          ]
        }
      ],
      "upper": []
    },
    {
      "variable": "over_a_0",
      "lower": [
        {
          "type_name": "(Double) => Double",
          "choices": [
            "a₃"
          ]
        },
        {
          "type_name": "(Int) => Int",
          "choices": [
            "a₁"
          ]
        },
        {
          "type_name": "(String) => String",
          "choices": [
            "a₂"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "(A_2) => B_3",
          "choices": []
        }
      ]
    },
    {
      "variable": "x_1",
      "lower": [],
      "upper": [
        {
          "type_name": "A_2",
          "choices": []
        }
      ]
    }
  ],
  "errors": []
}
